### ISSUE TITLE
String: New `Get Regexp Matches` keyword

### ISSUE THREAD

**Issue by synsun** [robotframework/robotframework#1985]
I want find multiple match result in a regexp. But 'Should Match Regexp' does not support multiple result.

``` python
import re
p = re.compile(ur'(?P<name>a)b', re.MULTILINE)
test_str = u"abcdeabababbbaaba"

re.findall(p, test_str)
```

```
${result}=  Should Match Regexp abcdeabababbbaababc (?P<name>ab)c
log ${result}       

INFO : ${result} = [u'abc', u'ab']
```

I don't know what is the good form to get multiple named result. (maybe ['abc','ab','abc','ab'] or [['abc','ab'],['abc','ab']])
I think add an option is good to apply that. Thanks.

**Comment by pekkaklarck** [comment]
Do you have a concrete idea how to enhance `Should Match Regexp` without breaking backwards compatibility?

**Comment by MoonSungHoon** [comment]
```
Should Match Regexp | ${string} | ${regexp} | global=false
```
If global option is true, return all matches. ( Don't return on first matches. )
Python regexp default value is 'global=on'. But not to break backwards compatibility, above option should be needed.

**Comment by MoonSungHoon** [comment]
Suggest new built-in keyword like below.
```
    def get_matches_regexp(self, string, regexp, group_name=None):
        """Returns match result of regexp

        Example:
        | ${results} =         | Get Matches Group Regexp | ${string}         | ${regexp} | ${group_name} |
        """
        import re
        c = re.compile(regexp, re.MULTILINE)
        if group_name:
            m = c.match(string)
            results = [m.groupdict()[group_name] for m in c.finditer(string)]
        else:
            results = [''.join(m) for m in c.findall(string)]
        return results
```

**Comment by pekkaklarck** [comment]
Could you explain, preferably with examples, how that global option would change the behavior? How would the implementation look like?

What is the use case for the new keyword you propose? Is that somehow related to the earlier enhancement you proposed?

**Comment by MoonSungHoon** [comment]
Okay, the new keyword is another way to get regexp results with global option.

This is an example why global option is needed.
I want find some longitude and latitude in a big text messages. That include some kind of texts and longitude, latitude ( I cannot define how many longitude, latitude ). After that using results of regexp filtering it using conditions to printout.

(Below text is just an example, not actual value)
```
Amenu    Bmenu    Logitude    Latitude
amanda  safe         W'114.54   N'36.10
bob         unsafe     W'401.41   N'61.20
jay          safe         W'102.11   N'12.11

Step1. Get logitude and latitude list = [114.54, 401.41, 102.11], [36.10, 61.20, 12.11]
Step2. filter that values ( over 120 at longitude, over 50 at latitude )
Step3. Get results 401.41, 61.20
```

**Comment by pekkaklarck** [comment]
I would still like to get a concrete example using the new keyword or enhanced old keyword. An example why the current functionality doesn't work would be good too.

**Comment by MoonSungHoon** [comment]
Why the current functionality doesn't work?
-> There are no keyword getting match results at a time. If I try to get those things using current function, it could be done but it is more complex and difficult.

More than actual example, I can explain an example. I want make a stastics using big log file containing large number of same format string. I want know the values in a format string satisfy a condition(find using regexp). How do I do that using current keywords?

**Comment by pekkaklarck** [comment]
Sorry, I don't fully understand the example you explained. I'd like to see a concrete example in Robot format that shows how the new features you propose actually work.

**Comment by MoonSungHoon** [comment]
This is a simple example for this. I want get INVOCATION_ID which STATE is TEST_OK.
```
${result} =
   -----------------------------------
   INVOCATION_ID   STATE              
   -----------------------------------
   0               TEST_OK 
   1               TEST_NOT_OK
   2               TEST_OK
   3               TEST_OK
   4               TEST_NOT_OK
   -----------------------------------
   COUNT = 2
;
```
And this is a python code for what I want to do.
```
import re
p = re.compile(ur'(\d+)\s+TEST_OK')
test_str = u"   -----------------------------------\n   INVOCATION_ID   STATE              \n   -----------------------------------\n   0               TEST_OK \n   1               TEST_NOT_OK\n   2               TEST_OK\n   3               TEST_OK\n   4               TEST_NOT_OK\n   -----------------------------------\n   COUNT = 2\n;"
 
re.findall(p, test_str)
```
And this is a robot framework example using new keyword.
```
*** Test Cases ***
aaa
    ${result}    Set Variable     \ \ -----------------------------------\\n \ \ INVOCATION_ID \ \ STATE \ \ \ \ \ \ \ \ \ \ \ \ \ \\n \ \ -----------------------------------\\n \ \ 0 \ \ \ \ \ \ \ \ \ \ \ \ \ \ TEST_OK \\n \ \ 1 \ \ \ \ \ \ \ \ \ \ \ \ \ \ TEST_NOT_OK\\n \ \ 2 \ \ \ \ \ \ \ \ \ \ \ \ \ \ TEST_OK\\n \ \ 3 \ \ \ \ \ \ \ \ \ \ \ \ \ \ TEST_OK\\n \ \ 4 \ \ \ \ \ \ \ \ \ \ \ \ \ \ TEST_NOT_OK\\n \ \ -----------------------------------\\n \ \ COUNT = 2\\n;
    ${ok}    Get Matches Regexp    ${result}    (?P<ID>\\d+)\\s+TEST_OK    ID
    log    ${ok}

INFO : ${ok} = [u'0', u'2', u'3']
```
If there are a lot of results like this(over 100!), how do I handle this?

**Comment by pekkaklarck** [comment]
Ok, I think I got the idea and can see this kind of keyword being useful. I don't think the keyword is generic enough to belong to BuiltIn library, but it could be added to String. Also I think `Get Regexp Matches` would be a better name than `Get Matches Regexp`.

Are you @MoonSungHoon interested in providing a pull request with code, tests and documentation? If yes, I'm happy to review it and merge it when it is ready. One of the first things to agree would be deciding what arguments the keyword should accept.

**Comment by MoonSungHoon** [comment]
Sure, I already made this keyword but this is my first time that commit to open source project. So please check my pull request. Im not familiar with contribute actually. :) Thank you.

**Comment by MoonSungHoon** [comment]
@pekkaklarck, please check my pull request. :)

**Comment by pekkaklarck** [comment]
Sorry for neglecting this issue for so long. I have been very busy with getting other higher priority issues ready for RF 2.9. There's still time to get this one into it too if you are interested.

On a quick look the PR looks pretty good and the keyword useful. There are some documentation issues but I can easily fix them myself. I would like to think about the proposed arguments and functionality of the keyword, though.

Being able to get only a certain group sounds like a useful feature, but why is is restricted for named groups? Wouldn't it also be handy to be able to give groups as an index? Additionally, should we support returning multiple groups in one go? Basically, could the implementation be something like this:

```python
def get_regexp_matches(string, pattern, *groups):
    regexp = re.compile(pattern)
    groups = [_parse_group(g) for g in groups]
    return [m.group(*groups) for m in regexp.finditer(string)]

def _parse_group(group):
    try:
        return int(group)
    except ValueError:
        return group
```

**Comment by MoonSungHoon** [comment]
Thank you for reply. I think your suggestion looks great.
Because I just want pick a specific result, I couldn't thought about that.
But multiple groups in one go would be better. :)

**Comment by pekkaklarck** [comment]
Great! Are you interested in updating your PR to match that kind of implementation? All your old tests ought to pass with this implementation, but some new should obviously be added and docs updated. We are a bit in a hurry to get this into RF 2.9 that's still possible. We try to get the first release candidate out at the end of next week.

**Comment by MoonSungHoon** [comment]
I change my pull request to apply your suggestion, @pekkaklarck

**Comment by pekkaklarck** [comment]
Could you add examples to docs as well as tests for using more that one index or group name?

**Comment by pekkaklarck** [comment]
A problem with using more than one group index or name is that it isn't that easy to operate the returned list of tuples in Robot Framework data. In Python you can just do

```python
>>> matches = [match.group(1, 2) for match in re.finditer('(a)(.)', 'a1a2...aN')]
>>> for first, second in matches:
...     print first, second
```

but in Robot data you need to do this:

```robotframework
*** Keywords ***
Example
    @{matches} =    Get Regexp Matches    a1a2...aN    (a)(.)    1    2
    :FOR    ${match}    IN    @{matches}
    \    ${first}    ${second} =    Set Variable    ${match}
    \    Log Many    ${first}    ${second}
```

It would be more convenient to just use

```robotframework
*** Keywords ***
Example
    @{matches} =    Get Regexp Matches    a1a2...aN    (a)(.)    1    2
    :FOR    ${first}    ${second}    IN     @{matches}
    \    Log Many    ${first}    ${second}
```

but that would require changing how for loops with more than one variable is handled in backwards incompatible manner. Does anyone have any idea how much people are using that functionality and is the current behavior considered useful? Any comments from @edbrannin who has recently been involved with for loops?

Regardless would it be a good idea to change for loops, I still think `Get Regexp Matches` accepting and returning multiple groups is a good idea. It's good that when there's exactly one group, Python automatically returns just a list of groups, not a list of tuples each containing just one group.

**Comment by pekkaklarck** [comment]
This feature looks useful enough and PR #2049 is already in so good shape I added this tentatively to RF 2.9. I hope this would be ready by Wednesday when we try to get beta 2 out. The final deadline is early next week when we plan to have the first release candidate and will start a feature freeze.

**Comment by edbrannin** [comment]
@pekkaklarck I think your last example ("more convenient to just use") would be a good candidate for a For-in-zip loop:

```
*** Keywords ***
Example
    @{matches} =    Get Regexp Matches    a1a2...aN    (a)(.)    1    2
    : FOR    ${first}    ${second}    IN ZIP    @{matches}
    \    Log Many    ${first}    ${second}
```

(I haven't executed the above and need to go AFK now, but I think it should work.)

**Comment by MoonSungHoon** [comment]
@edbrannin I'm on test this keyword with multiple groups but, "IN ZIP" does not work.

**Comment by edbrannin** [comment]
@MoonSungHoon By "Does not work", do you mean (A) you're working with a version of Robot from before my for-in-zip PR was merged (in the past 2 days), or (B) some other error message (which?)

**Comment by MoonSungHoon** [comment]
@edbrannin (A) Oh, I didn't know that it is a new feature. I'll update my version. Thanks.

**Comment by MoonSungHoon** [comment]
To use "IN ZIP" first and second results should be enumerated in each array, but get_regexp_matches return combined array like [('1','2'), ('1','2')]. I think extend "IN" like python code would be great.

```
*** Keywords ***
Example
    @{matches} =    Get Regexp Matches    a1a2...aN    (a)(.)    1    2
    : FOR    ${first}    ${second}    IN ZIP    @{matches}
    \    Log Many    ${first}    ${second}
```
This test case return FAILED with "FOR IN ZIP expects an equal number of variables and iterables. Got 2 variables and 1 iterable.

**Comment by MoonSungHoon** [comment]
@pekkaklarck I pushed with additional docs for groups as name and index and add more atest.

**Comment by pekkaklarck** [comment]
Yeah, for-in-zip loops don't unfortunately work in this case. It would be better if normal for loop would work, like it does in Python, but I'm not certain is changing syntax worth the effort. Adding a new for loop flavor would be possible too, but that might also just make the whole for loop usage more complicated.

@MoonSungHoon, I added few more comments to the PR. Could you take a look at them? It would also be great if you could squash commits to make the history a little cleaner. There shouldn't bee too many conflicts so doing that should be pretty simple. If you are new to squashing, here are pretty good instructions:
http://eli.thegreenplace.net/2014/02/19/squashing-github-pull-requests-into-a-single-commit

**Comment by edbrannin** [comment]
I was looking at our internal `HelperLibrary.py` yesterday and found a keyword that's perfect for this situation, at the expense of one extra keyword:

It went something like this:

```python
def zip_and_flatten(iterable):
    answer = list()
    for item in zip(*iterable):
        answer += item
    return answer
```

```robot
*** Settings ***
Library    zip_and_flatten.py

*** Test Cases ***

Test 1985
    ${matches} =    Evaluate    [(2,2), (3,4)]
    ${matches}=    Zip and Flatten    ${matches}
    : FOR    ${first}    ${second}    IN     @{matches}
    \    Log Many    ${first}    ${second}
```

**Comment by MoonSungHoon** [comment]
@pekkaklarck, I applied your comments on PR.
But I gave up squashing because it was too hard to merge and rebase with commits.
( How do I make a merged new single commit on already mixed branch? )
So I made a new PR. Please check that.

**Comment by pekkaklarck** [comment]
I merged PR #2052 and then cleaned up tests a little in 600aaeebd308ad9f1c8e3e2b55cef628b51f3146 and enhanced docs in 78abd3235e24d43f9a2957022f601d22b305d8ad. This issue can be considered done and closed.

Thanks for submitting the issue and the PR @MoonSungHoon! Please take a look my changes, especially docs, and comment here if you see something that should still be improved.
